### DISCUSSION TITLE
JsonResource With pagination feature

### DISCUSSION THREAD

**Discussion by hsmfawaz** [laravel/framework#37837] [Ideas]
Hello Laravel Team ,
I wanna  introduce to you my idea to enhance JsonResource class as the following 

**The problem:**

For the last 3 years ,
I was working on providing API for mobile applications using Laravel and Lumen 
using laravel JsonResource helped me a lot to make a good api responses, until now no problem,

The problem come with Pagination when i want to paginate a JsonResource ,I create a JsonCollection to  handle the pagination.

at the end of every application ,I end up with a ton of JsonCollection has the same functionality and 99% same code  **only the resource name is different** 

Something like this:

```PHP
class AdsCollection extends ResourceCollection
{
    public $collects = AdResource::class;

    public function toArray($request)
    {
        return [
            'ads' => $this->collection,
            'paginate' => [
                'total' => $this->total(),
                'count' => $this->count(),
                'per_page' => $this->perPage(),
                'next_page_url' => $this->nextPageUrl() ?? '',
                'prev_page_url' => $this->previousPageUrl() ?? '',
                'current_page' => $this->currentPage(),
                'total_pages' => $this->lastPage(),
            ],
        ];
    }
}

```

The solution:
I tried many solutions to eliminate the unneeded collections files
My last solution was implementing a HasPagination Trait and use it in JsonResource

//HasPagination Trait
```PHP
trait HasPagination
{
    public static function paginate($resource, $wrapper = 'items')
    {
        if ($resource instanceof AbstractPaginator) {
            return new class($resource, self::class, $wrapper) extends ResourceCollection{
                public string $wrapper;

                public function __construct($resource, string $collects, $wrapper = 'items')
                {
                    $this->collects = $collects;
                    parent::__construct($resource);
                    $this->wrapper = $wrapper;
                }

                public function toArray($request) : array
                {
                    return [
                        $this->wrapper => $this->collection,
                        'paginate'     => [
                            'total'         => $this->total(),
                            'count'         => $this->count(),
                            'per_page'      => $this->perPage(),
                            'next_page_url' => $this->nextPageUrl() ?? '',
                            'prev_page_url' => $this->previousPageUrl() ?? '',
                            'current_page'  => $this->currentPage(),
                            'total_pages'   => $this->lastPage(),
                        ],
                    ];
                }
            };
        }

        return self::collection($resource);
    }
}
```
 Now i can paginate JsonResource like
```PHP
$collection = Ad::latest()->paginate();
return AdResource::paginate($collection);
```

I will enhance the trait to be able to paginate cursor pagination and simple pagination,
Maybe enhancing the static collection method in the JsonResource to handle the pagination along side the collection
too many enhancements in my head to handle the pagination functionality with the jsonResource 
What is your opinion about my idea ?

Thanks so much for your time.

**Comment by jelib3an** [comment]
Hi, I was reading the laravel docs and it appears resource collections already work with pagination. although I haven't used it, but does that work for what you are trying to accomplish?
https://laravel.com/docs/8.x/eloquent-resources#pagination

**Reply by hsmfawaz** [reply]
Hello jelib3an ,Yes i know that resource collection (JsonCollection) has the pagination feature.
But using the JsonCollection for the pagination feature only will produce too many jsonCollection doing the same functionality as i described.

**Reply by jelib3an** [reply]
I see. Since you mentioned trying many solutions before settling on using a trait, curious to know, what else have you tried?

**Reply by hsmfawaz** [reply]
I created a ResourceCollection then override the __construct to be able to  accept JsonResource 

```PHP
class BaseCollection extends ResourceCollection
{
    public function __construct($resource, string $collects)
    {
        $this->collects = $collects;
        parent::__construct($resource);
    }

```

**Comment by FalkHe** [comment]
Why not simply extend the ResourceCollection to your needs?   IMO pagination information should not be in `toArray()` but in `paginationInformation()`

That's what I do:

```php
class AbstractJsonApiCollection extends ResourceCollection
{
    public function paginationInformation($request, $paginated, $default): array
    {
        return [
            'links' => $default['links'],
            'meta' => [
                'firstItemNumber' => $default['meta']['from'],
                'lastItemNumber' => $default['meta']['to'],
                'totalItemCount' => $default['meta']['total'],
                'itemsPerPage' => $default['meta']['per_page'],
                'currentPageNumber' => $default['meta']['current_page'],
                'lastPageNumber' => $default['meta']['last_page']
            ]
        ];
    }
}
``´

this way, you only need to extend your Collection from the Abstract one. At least no duplicate code.

```php
class MyCollection extends AbstractJsonApiCollection {}
```

that's all

**Reply by hsmfawaz** [reply]
@FalkHe The main idea is not to create a ResourceCollection and just the JsonResource alone

**Reply by FalkHe** [reply]
May be, i don't get it, but: 

You can do even this by overwriting `newCollection()` in your JsonResource and use your customized Collection instead of the `AnonymousResourceCollection`.

**Comment by macropay-solutions** [comment]
@hsmfawaz you can have a look here as an alternative: https://github.com/macropay-solutions/laravel-crud-wizard-free
and at the demo page http://89.40.19.34/laravel-10-free/laravel-lumen-crud-wizard

It supports length aware paginator, simple paginator and cursor. You also can overwrite the controller function https://github.com/macropay-solutions/laravel-crud-wizard-free/blob/1d3a509651ad6be1b2742445ffe9181187b8dce0/src/Http/Controllers/ResourceControllerTrait.php#L275 to format the paginator fields in the response.

**Comment by murilolivorato** [comment]
I do like this -

```

namespace App\Http\Resources\Post;

use Illuminate\Http\Request;
use Illuminate\Http\Resources\Json\JsonResource;

class PostPaginateResource extends JsonResource
{
    /**
     * Transform the resource into an array.
     *
     * @return array<string, mixed>
     */
    public function toArray(Request $request): array
    {
        return [
            "data" => PostResource::collection($this->items()),
            'pagination' => [
                'total'         => $this->total(),
                'per_page'      => $this->perPage(),
                'current_page'  => $this->currentPage(),
                'last_page'     => $this->lastPage(),
                'from'          => $this->firstItem(),
                'to'            => $this->lastItem()
            ]
        ];
    }

}

```
